src/rust/engine/fs/store/src/lib.rs

```diff
@@ -468,93 +481,58 @@ impl Store {
-          return future::ok((ingested_digests.keys().cloned().collect(), ingested_digests))
-            .to_boxed();
-        }
-        let request = remote.find_missing_blobs_request(ingested_digests.keys());
-        let f = remote.list_missing_digests(request);
-        f.map(move |digests_to_upload| (digests_to_upload, ingested_digests))
-          .to_boxed()
-      })
-      .and_then(move |(digests_to_upload, ingested_digests)| {
-        future::join_all(
-          digests_to_upload
-            .into_iter()
-            .map(|digest| {
-              let entry_type = ingested_digests[&digest];
-              let local = local.clone();
-              let remote = remote2.clone();
-
-              Box::pin(async move {
-                let executor = local.executor().clone();
-                let maybe_upload = local
-                  .load_bytes_with(entry_type, digest, move |bytes| {
-                    // NB: `load_bytes_with` runs on a spawned thread which we can safely block.
-                    executor.block_on(remote.store_bytes(bytes))
-                  })
-                  .await?;
-                match maybe_upload {
-                  Some(future) => Ok(future),
-                  None => Err(format!("Failed to upload digest {:?}: Not found", digest)),
-                }
-              })
-              .compat()
-              .to_boxed()
-            })
-            .collect::<Vec<_>>(),
-        )
-        .and_then(future::join_all)
-        .map(|uploaded_digests| (uploaded_digests, ingested_digests))
-      })
-      .map(move |(uploaded_digests, ingested_digests)| {
-        let ingested_file_sizes = ingested_digests.iter().map(|(digest, _)| digest.1);
-        let uploaded_file_sizes = uploaded_digests.iter().map(|digest| digest.1);
-
-        UploadSummary {
-          ingested_file_count: ingested_file_sizes.len(),
-          ingested_file_bytes: ingested_file_sizes.sum(),
-          uploaded_file_count: uploaded_file_sizes.len(),
-          uploaded_file_bytes: uploaded_file_sizes.sum(),
-          upload_wall_time: start_time.elapsed(),
-        }
+          ingested_digests.keys().cloned().collect()
+        } else {
+          let request = remote.find_missing_blobs_request(ingested_digests.keys());
+          remote.list_missing_digests(request).compat().await?
+        };
+
+      let uploaded_digests = future03::try_join_all(
+        digests_to_upload
+          .into_iter()
+          .map(|digest| {
+            let entry_type = ingested_digests[&digest];
+            let local = store.local.clone();
+            let remote = remote.clone();
+
+            async move {
+              let executor = local.executor().clone();
+              let maybe_upload = local
+                .load_bytes_with(entry_type, digest, move |bytes| {
+                  // NB: `load_bytes_with` runs on a spawned thread which we can safely block.
```

Comment:
This seems like an assumption that could confuse a reader (if not for the comment). Should `load_bytes_with` be renamed to say it can block (e.g., `load_bytes_with_blocking`)? Have all other load_bytes_with call sites received a similar comment? Does Rust have a way to encode the "blockability" of a closure into the type? (probably not but I'm curious if it has)

Reply:
> This seems like an assumption that could confuse a reader (if not for the comment). Should load_bytes_with be renamed to say it can block (e.g., load_bytes_with_blocking)?

This is existing code that was only moved: see #9793. The called method also has a comment.

> Does Rust have a way to encode the "blockability" of a closure into the type? (probably not but I'm curious if it has)

Good question: whee, colored functions.

No, it does not. On the "bright side", tokio will panic to let you know that you've blocked in the wrong place: so you'll know if you've done it wrong, but you don't get compiler assistance.

Reply:
👍